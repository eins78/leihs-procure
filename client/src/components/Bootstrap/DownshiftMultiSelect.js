import React from 'react'
import cx from 'classnames'
// import Downshift from 'downshift'

// NOTE: work in progress, right now only "vanilla DOM" with option groups.
//       props/API will stay the same even if implemented using Downshift.

// import ButtonDropdown from 'reactstrap/lib/ButtonDropdown'
// import DropdownToggle from 'reactstrap/lib/DropdownToggle'
// import DropdownMenu from 'reactstrap/lib/DropdownMenu'
// import DropdownItem from 'reactstrap/lib/DropdownItem'

import { getSelectedValueFromProps, getSelectedValueFromEvent } from './Select'

const MultiSelect = ({
  options,
  value,
  onChange,
  readOnly,
  multiple = true,
  className,
  ...props
}) => {
  if (!multiple) {
    throw new Error('MultiSelect only supports `multiple=true`!')
  }

  if (onChange && !props.name) {
    throw new Error('Input with `onChange` needs a `name`!')
  }

  const selectedValue = getSelectedValueFromProps(value, multiple)

  const cls = cx('form-control', className)

  return (
    <select
      multiple
      value={selectedValue}
      className={cls}
      disabled={readOnly}
      onChange={e => {
        if (props.readOnly) return e.preventDefault()
        onChange({
          target: {
            name: props.name,
            value: getSelectedValueFromEvent(e, multiple)
          }
        })
      }}
      {...props}
    >
      {options.map(({ label, options }, i) => (
        <optgroup key={i} label={label}>
          {options.map(({ label, ...props }, i) => (
            <option key={i} {...props}>
              {label}
            </option>
          ))}
        </optgroup>
      ))}
    </select>
  )
}

export default MultiSelect

// function titleBySelection(selected, values) {
//   const count = selected.length
//   const valuesCount = values.length
//   if (count === valuesCount) {
//     return 'Alle ausgew\xE4hlt'
//   }
//   if (count === 0) {
//     return 'Keine ausgew\xE4hlt'
//   }
//   if (count > 3) {
//     return `${count} selektiert`
//   }
//   return selected
//     .map(item => values.filter(({ value }) => value === item)[0])
//     .map(item => item && item.label)
//     .join(', ')
// }
