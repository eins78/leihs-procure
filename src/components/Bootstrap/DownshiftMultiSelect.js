import React from 'react'
// import Downshift from 'downshift'

// import ButtonDropdown from 'reactstrap/lib/ButtonDropdown'
// import DropdownToggle from 'reactstrap/lib/DropdownToggle'
// import DropdownMenu from 'reactstrap/lib/DropdownMenu'
// import DropdownItem from 'reactstrap/lib/DropdownItem'

const MultiSelect = ({ options, ...rest }) => {
  return (
    <select multiple {...rest}>
      {options.map(({ label, options }, i) => (
        <optgroup key={i} label={label}>
          {options.map(({ label, ...props }, i) => (
            <option key={i} {...props}>
              {label}
            </option>
          ))}
        </optgroup>
      ))}
    </select>
  )
}

export default MultiSelect

// function titleBySelection(selected, values) {
//   const count = selected.length
//   const valuesCount = values.length
//   if (count === valuesCount) {
//     return 'Alle ausgew\xE4hlt'
//   }
//   if (count === 0) {
//     return 'Keine ausgew\xE4hlt'
//   }
//   if (count > 3) {
//     return `${count} selektiert`
//   }
//   return selected
//     .map(item => values.filter(({ value }) => value === item)[0])
//     .map(item => item && item.label)
//     .join(', ')
// }
