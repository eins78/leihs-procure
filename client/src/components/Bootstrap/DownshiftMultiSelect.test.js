// import React from 'react'
import renderer from 'react-test-renderer'

import { examples } from './DownshiftMultiSelect.examples'

examples.forEach(({ name, content }) => {
  it(`${name}: renders correctly`, () => {
    expect(renderer.create(content).toJSON()).toMatchSnapshot()
  })
})
