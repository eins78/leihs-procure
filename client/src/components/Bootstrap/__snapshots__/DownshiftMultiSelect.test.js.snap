// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelect: renders correctly 1`] = `
<div>
  <select
    className="form-control"
    disabled={undefined}
    id="example-multiple-optgroups"
    multiple={true}
    onChange={[Function]}
    value={
      Array [
        undefined,
      ]
    }
  >
    <optgroup
      label="Group 1"
    >
      <option
        value="1-1"
      >
        Option 1.1
      </option>
      <option
        selected={true}
        value="1-2"
      >
        Option 1.2
      </option>
      <option
        selected={true}
        value="1-3"
      >
        Option 1.3
      </option>
    </optgroup>
    <optgroup
      label="Group 2"
    >
      <option
        value="2-1"
      >
        Option 2.1
      </option>
      <option
        value="2-2"
      >
        Option 2.2
      </option>
      <option
        value="2-3"
      >
        Option 2.3
      </option>
    </optgroup>
  </select>
</div>
`;
